### ISSUE TITLE
tribits_add_test does not support generator expressions for properties

### ISSUE THREAD

**Issue by jjwilke** [trilinos/Trilinos#8110]
When using `tribits_add_test`, we are unable to set test properties afterwards using generator expressions.

````
set_property(TEST ${TEST_NAME}  APPEND_STRING PROPERTY ENVIRONMENT 
   "KOKKOS_PROFILE_LIBRARY=$<TARGET_FILE:${TEST_TOOL}>")
````

If I create tests using `add_test`, everything works. As soon as I go through `tribits_add_test`, I can no longer set test properties using generator expressions. I am totally baffled by this behavior as to what `tribits_add_test` could be doing that makes generator expressions not work.

**Comment by cgcgcg** [comment]
@bartlettroscoe

**Comment by bartlettroscoe** [comment]
@jjwilke, you sure you have the right test name and that the test is actually added?  tribits_add_test() just calls add_test()  and therefore you should be able to set any properties you like as described at:

* https://tribits.org/doc/TribitsDevelopersGuide.html#setting-additional-test-properties-tribits-add-test

And what is meant by "not works"?

**Comment by jjwilke** [comment]
Yes, definitely sure I have the right test name. Sorry - should have sent some more detail. We get the following error when trying to run the test:

````

**Comment by bartlettroscoe** [comment]
> Somehow the generator expression is being interpreted as literal text and not getting processed.  There has to be some policy or something set on how to interpret the $ character, but I can't fathom what that could be.

@jjellio, this should not be rocket science.  What does the generated `CTestTestfile.cmake` file look like in both cases?  We should be able to debug the generation of that file with some help from Kitware.

In any case, do you have a branch pushed somewhere so that we can reproduce this?

**Comment by jjwilke** [comment]
There is an open PR here: https://github.com/trilinos/Trilinos/pull/8108

**Comment by jjwilke** [comment]
The CTest file has the generator expression escaped "/$<".

**Comment by bartlettroscoe** [comment]
> The CTest file has the generator expression escaped "/$<".

@bradking, @zackgalbreath,

Any idea why the the generator expression would not work correctly in:

https://github.com/trilinos/Trilinos/pull/8108/files#diff-9bb184b36b679a3ad86de2516b9d30a8R116

?

Why would it be escaped in this way?

**Comment by bradking** [comment]
Test properties don't currently support generator expressions, IIRC.

**Comment by jjwilke** [comment]
@bradking 
The following minimal reproducer uses generator expressions and works. 
Output of `ctest --extra-verbose`:
````
test 1
    Start 1: thetest

1: Test command: /Users/jjwilke/Programs/cmake-tests/genexp/build/mytest
1: Environment variables: 
1:  MY_ENV=/Users/jjwilke/Programs/cmake-tests/genexp/build/mytest
1: Test timeout computed to be: 10000000
1/1 Test #1: thetest ..........................   Passed    0.01 sec
````

CMake code:
````
cmake_minimum_required(VERSION 3.12)
project(test CXX)
add_executable(mytest test.cc)
enable_testing()
add_test(NAME thetest COMMAND mytest)
set_property(TEST thetest APPEND_STRING
             PROPERTY ENVIRONMENT
             MY_ENV=$<TARGET_FILE:mytest>)
````

The test.cc file is basically empty:
````
int main(){ return 0; }
````

The CMake documentation says: 
" Generator expressions will be expanded the same as supported by the test’s add_test() call."

**Comment by bradking** [comment]
Ah, now I remember that, thanks.  The `add_test(NAME ... COMMAND ...)` form will expand generator expressions, but the plain `add_test(<name> <command>...)` form will not.

**Comment by bartlettroscoe** [comment]
> Ah, now I remember that, thanks.  The `add_test(NAME ... COMMAND ...)` form will expand generator expressions, but the plain `add_test(<name> <command>...)` form will not.

@bradking, when did CMake support the `add_test(NAME ... COMMAND ...)` form?  What version of CMake?

**Comment by bradking** [comment]
The `add_test(NAME ... COMMAND ...)` form was added in CMake 2.8.0, circa 2009.

**Comment by jjwilke** [comment]
Yeeeppppp. That does it. Sorry, I should have seen in the add_test documentation. 

@bartlettroscoe - what do you think?
````cmake
FUNCTION(TRIBITS_ADD_TEST_ADD_TEST TEST_NAME EXE_NAME)
  IF (NOT TRIBITS_ADD_TEST_ADD_TEST_UNITTEST)
    ADD_TEST(${TEST_NAME} ${ARGN})
  ENDIF()
  TRIBITS_SET_TESTS_PROPERTIES(${TEST_NAME} PROPERTIES REQUIRED_FILES ${EXE_NAME})
ENDFUNCTION()
````

would change to:

````cmake
FUNCTION(TRIBITS_ADD_TEST_ADD_TEST TEST_NAME EXE_NAME CMD)
  IF (NOT TRIBITS_ADD_TEST_ADD_TEST_UNITTEST)
    ADD_TEST(NAME ${TEST_NAME} COMMAND ${ARGN})
  ENDIF()
  TRIBITS_SET_TESTS_PROPERTIES(${TEST_NAME} PROPERTIES REQUIRED_FILES ${EXE_NAME})
ENDFUNCTION()
````

**Comment by jjwilke** [comment]
and if so - should that be merged as a separate PR?

**Comment by bartlettroscoe** [comment]
@jjwilke, I will make this change in TriBITS (make sure all the TriBITS test still pass) and then post a PR in Trilinos to merge in the updated TriBITS snaphsot.

**Comment by bartlettroscoe** [comment]
FYI: Interestingly, when you switch from:

```
add_test(<name> <command>)
```

to

```
add_test(NAME <name> COMMAND <command>)
```

CMake will assert at configure time when there are more than one test in the same directory.  (I am amazed that this was not asserted before.)

There was actually a duplicate test in the TriBITS test suite (from a copy and paste at some point years ago).

This means that making this change in TriBITS is going to break backwards compatibility.  I would guess this is going to flag some duplicate tests in Trilinos as well when I try to apply this to Trilinos. 

@bradking, what is the behavior of CMake/CTest when defining duplicate tests with:

```
add_test(<name> <command>)
```

Does it just use the last test that was defined or does it do something else?  (Or is it undefined?)

**Comment by bradking** [comment]
With the old style I think the test is duplicated, though this is not documented or tested.  When `add_test(NAME)` was added, the intention was to resolve such problems in the new signature without breaking compatibility on clients using the old signature.

**Comment by bartlettroscoe** [comment]
> With the old style I think the test is duplicated, though this is not documented or tested

@bradking, but if you set a test property after both tests were defined with add_test(), what test got the property?

**Comment by bradking** [comment]
Oops, I misread the old behavior.  Adding a test multiple times with the old signature is only recorded the first time.

**Comment by bartlettroscoe** [comment]
> The `add_test(NAME ... COMMAND ...)` form was added in CMake 2.8.0, circa 2009.

TriBITS was first written in 2008 against CMake 2.4.0.  That explains the form of `add_test()` in TriBITS to this point.  (I guess this is the first case where it really mattered.)

**Comment by bartlettroscoe** [comment]
> Oops, I misread the old behavior. Adding a test multiple times with the old signature is only recorded the first time.

Okay, to resolve duplicate tests in Trilinos, I will keep the first test and delete the following duplcates.

**Comment by bartlettroscoe** [comment]
@jjwilke, the PR #8115 should fully address this issue.  Can you please approve that PR?  After that, the full Kokkos tests suite should be able to be added and run as part of Trilinos.

**Comment by bartlettroscoe** [comment]
This should be resolved as of the merge of PR #8115 

Closing as complete.
